### DISCUSSION TITLE
Mark fields as type-identifiers

### DISCUSSION THREAD

**Discussion by rsinohara** [colinhacks/zod#594] [Ideas]
Would it make sense to set some fields as "type marker", meaning if those fields fail validation, no other fields need to be checked for that schema, moving on to the first parent union member?

We could then mark some fields, and avoid many errors that are due to checking the wrong types (and, when that happens, we end up looking for those fields anyway to see which type was the intended one).

Something like this:
```
{
   type: z.string().brand()
}
```

**Comment by colinhacks** [comment]
This would likely take the form of a higher-level `z.discriminatedUnion` construct. Something like:

```ts
z.discriminatedUnion([
  z.object({ kind: z.literal('a'), data1: z.string() }),
  z.object({ kind: z.literal('b'), data2: z.string() }),
  z.object({ kind: z.literal('c'), data3: z.string() }),
], 'kind')
```

With some static validation to make sure `'kind'` is a key shared by all elements of the array. Though I'm making this up off the cuff, open to other proposals by @scotttrinh or others.

**Reply by scotttrinh** [reply]
I've started working on this exact thing [here](https://github.com/zod-contrib/zod/tree/discriminated-union), and the benchmarks are looking 🔥. Definitely some work to do to make it more robust, but that's the exact API I've started with.

https://github.com/zod-contrib/zod/blob/81ce18e0a5678e31368b17fa53bff67077444b28/src/types.ts#L1811

**Reply by rsinohara** [reply]
Just a question, how would that work if the kind in nested, would that be possible? Or to have more than one?

**Reply by scotttrinh** [reply]
This leans into how TypeScript discriminated unions work, so I don't imagine that nesting or multiple fields is something that'd I'd think to support since TypeScript doesn't support them either.

Maybe there is some kind of construct like:

```ts
z.unionBy((input: FooUnion) => {
  if (input.kind === "FOO" && input.subkind === "BAR") {
    return fooBarSchema;
  }
  // ...etc
});
```

Where you specifying a way of discriminating at runtime and returning the member schema?

**Reply by scotttrinh** [reply]
FWIW, even if we had a way of validating the type at runtime with something like that, we wouldn't get TypeScript's support for it after parsing like we do with a simple discriminated union. Here is what I mean:

```ts
// Has compiler support:

const unionSchema = z.discriminatedUnion([
  z.object({ kind: z.literal("a"), data1: z.string() }),
  z.object({ kind: z.literal("b"), data2: z.number() }),
], "kind");

const data = unionSchema.parse(inputData);

switch (data.kind) {
  case "a": {
    // TS knows this is an "a" kind object
  }

  case "b": {
    // TS knows this is a "b" kind object
  }
}

// No compiler-support:

const runtimeUnionSchema = z.unionBy((input) => {
  if (input.nested.data.marker === "foo") {
    //etc
  }
  //etc
});

switch (/* what goes here? */) {
}
```

**Reply by rsinohara** [reply]
One thing I don't get. I was coming more from the error reporting angle, tagging a field as 'kind' to basically squelch all other errors if that field failed.

On the TS side, I see your point above, but won't TS be able to discriminate unions, even without the `discriminatedUnion` concept? It does that already no?

**Reply by scotttrinh** [reply]
Yeah, `z.union` definitely works to give you discriminated unions at compile-time already, but it runs all of the schemas and reports all of the errors. Which has both error-related ergonomic issues and performance implications: since it runs all of the schemas performance is linear to the number of all internal schema checks instead of just the checks in the "correct" schema.

See the benchmark output in this PR for some details: https://github.com/colinhacks/zod/pull/578

**Reply by rsinohara** [reply]
I have a few questions still. You can't point me to the right place to learn about discriminated unions, or a better place to discuss. We could also have a chat if you prefer.

But here it goes anyway.

The way I see it, this discrimination is more a property of a field, not the union itself (since TS discrimination is not a factor). Something like "if this field fails validation, don't bother with checking and outputting errors for other fields" (also probably checking those fields before others).
Meaning two schemas inside an union could have distinct discriminating fields (or, easily, sets of). `.kind()` on those fields only sets a flag with this behavior.


To me, the concept of kind seem like something that belongs to a schema, independently of it being part of an union or not.


PS: I am probably missing a lot of history/experience with zod, not so much challenging but trying to understand ;)

**Reply by scotttrinh** [reply]
> Something like "if this field fails validation, don't bother with checking and outputting errors for other fields" (also probably checking those fields before others).

This makes sense and as you've said, this is somewhat orthogonal to discriminating between different members of a union. If I understand correctly, this is more of a performance optimization since it's saying "check this/these field(s) first, and if they fail, don't continue checking". I can definitely see a use case for this, but I'm concerned that it would lead to confusion about how and when to use a feature like this. I _think_ the order of the keys in an object definition determines the order in which we check the values, so maybe having another parsing mode that fails on the first error rather than returning all errors might satisfy this use case and allow the schema _consumer_ to determine the behavior that they desire?

Something like —
```ts
someSchema.failFastParse(input);
```

**Reply by rsinohara** [reply]
I think that will also work.

My angle is more of error output, but of course in this case it goes hand in hand. What you just proposed would work too for me.

**Reply by rsinohara** [reply]
@scotttrinh Do you think it's something I should try implementing?

**Reply by scotttrinh** [reply]
I would love to get some input from other maintainers, too. I wonder if we would want/need to implement a throwing and safe version, or maybe pass options to `parse`/`safeParse` instead of having a separate method? @colinhacks any thoughts on this?
